cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBPartitionManager.java

```diff
@@ -0,0 +1,291 @@
+        DirUtils.deleteDirectoryContents(partitionDir);
+        partitions.remove(timeInterval.startMillis);
+        numPruned++;
+      }
+    }
+    return numPruned;
+  }
+
+  public LevelDBPartition getOrCreatePartition(long publishTime) throws IOException {
+    Long partitionKey = partitions.floorKey(publishTime);
+    if (partitionKey != null) {
+      LevelDBPartition partition = partitions.get(partitionKey);
+      if (partition.getEndTime() > publishTime) {
+        return partition;
+      }
+    }
+
+    synchronized (this) {
+      // Check again to make sure no new instance was being created while this thread is acquiring the lock
+      partitionKey = partitions.floorKey(publishTime);
+      LevelDBPartition partition;
+      if (partitionKey != null) {
+        partition = partitions.get(partitionKey);
+        if (partition.getEndTime() > publishTime) {
+          return partition;
+        }
+      }
+
+      partition = getOrCreateDBPartition(publishTime);
+      partitions.put(partition.getStartTime(), partition);
+
+      return partition;
+    }
+  }
+
+  /**
+   * Get existing partitions for publish times equal to or greater than the given timestamp.
+   */
+  public Collection<LevelDBPartition> getPartitions(long startTime) throws IOException {
+    if (initialized.compareAndSet(false, true)) {
+      initPartitions();
+    }
+    Long partitionStartTime = partitions.floorKey(startTime);
+    return partitionStartTime == null ? partitions.values() : partitions.tailMap(partitionStartTime, true).values();
+  }
+
+  /**
+   * Get the partitions responsible for data between the specified start and end times.
+   *
+   * @param startTime inclusive start time
+   * @param endTime inclusive end time
+   * @return partitions responsible for data between the specified start and end times
+   */
+  public Collection<LevelDBPartition> getPartitions(long startTime, long endTime) throws IOException {
+    if (initialized.compareAndSet(false, true)) {
+      initPartitions();
+    }
+    Long partitionEndTime = partitions.floorKey(endTime);
+    if (partitionEndTime == null) {
+      // partitionEndTime can only be null if the partitions map is empty, or the endTime is less than the
+      // start time for any existing partition. In either case, there are no partitions that contain data
+      // published between the start and end times.
+      return Collections.emptyList();
+    }
+
+    Long partitionStartTime = partitions.floorKey(startTime);
+    // partitionStartTime can be null if the startTime is less than the start time for any existing partition
+    partitionStartTime = partitionStartTime == null ? 0 : partitionStartTime;
+    return partitions.subMap(partitionStartTime, true, partitionEndTime, true).values();
+  }
+
+  public synchronized void close() throws IOException {
+    IOException failure = null;
+    for (LevelDBPartition partition : partitions.values()) {
+      try {
+        partition.getLevelDB().close();
```

Comment:
A bit concerning in looping and closing in synchronized block. Better to take a copy of the partition values and clear the map in a sync block, then loop and close outside the sync block.